### PR TITLE
Invoke Request callback with error if connection closed prematurely.

### PR DESCRIPTION
**Pull request by tvrprasad**
This addresses https://github.com/tediousjs/tedious/issues/482

### CODE REVIEW

**Comment by tvrprasad** [comment]
If someone wrote code like this, they will now get request callback with error where they didn't before:
```JavaScript
  // Any one of 'done', 'doneInProc' and 'doneProc'
  request.on('done', function (rowCount, more, rows) {
    connection.close();
  });
```
This is invalid code as TDS allows for multiple of these tokens and request is not complete till the callback passed in constructor is invoked.

I think we should not consider this a API breaking change, but calling it out nevertheless in case there is a difference of opinion on this.

**Comment by arthurschreiber** [comment]
> This is invalid code as TDS allows for multiple of these tokens and request is not complete till the callback passed in constructor is invoked.

I believe that should be okay.

File: src/connection.js [hunk 1]
```diff
@@ -245,6 +245,11 @@ class Connection extends EventEmitter {
       } else {
         this.emit('rerouting');
       }
+      if (this.request) {
+        const err = RequestError('Connection closed before request completed.', 'ECLOSE');
```
Comment (arthurschreiber) [inline src/connection.js]: Could you change this to use `new` here?

```js
const err = new RequestError('Connection closed before request completed.', 'ECLOSE');
```

Comment (tvrprasad) [inline src/connection.js]: Thanks for merging the change. Looks like we're using the non-new form of RequestError everywhere, so leaving as is.

**Comment by neoadventist** [comment]
@tvrprasad @arthurschreiber hey guys, yeah this is breaking my application. any guidance on how to fix? 

```
   30 module.exports = {
   29         query:function(str){
   28                 return new Promise(function (resolve, reject) {
   27                         var connection  = new Connection(config);
   26                         connection.on('connect', function(err) {
   25                                 if(err){
   24                                         return reject(err);
   23                                 }
   22                                 var request = new Request(str, function(err, rowCount) {
   21                                         if (err) {
   20                                                 console.log(err);
   19                                                 return reject(err);
   18                                         } else {
   17                                                 console.log(rowCount + ' rows');
   16                                         }
   15                                 });
   14                                 request.on('done',function (rowCount, more, rows) {
   13                                         rows = rows.map(function(row){
   12                                                 return Object.keys(row).reduce(function(obj,key){
   11                                                         obj[key] = row[key].value;
   10                                                         return obj;
    9                                                 },{});
    8                                         });
    7                                         connection.close();
    6                                         return resolve(rows);
    5                                 });
    4                                 connection.execSqlBatch(request);
    3                         });
    2                 });
    1         }
  38  };
```

removing line seven seems to work, I just would like to know if not explicitly closing the connection will break anything. 

Thanks!

**Comment by tvrprasad** [comment]
@neoadventist My comment up above speaks to this specific thing :-) You should not close the connection in 'done' callback, instead you should do it in Request callback. You should move line 7 to line 15.

**Comment by neoadventist** [comment]
@tvrprasad thanks!

**Pull request merged**